Revise and reorganize "Why the Component Model?" section (#281)
bytecodealliance/component-docs: component-model/src/design/why-component-model.md

# Why the Component Model?

> WebAssembly programs can be written by hand,

Author: I see your point -- see what you think about [9f86658](https://github.com/bytecodealliance/component-docs/pull/281/commits/9f86658d3e0e237cdae8e5c499db928913fd45b9)?

The original text does start with quite a bit of explanation about core modules as well. By defining core modules first and then moving to components, I was hoping to avoid having to add a connecting sentence like "To understand what the limitations of core modules are, we start by defining them." But I ended up doing exactly that. I don't like this kind of sentence, but maybe you can think of a way to avoid it.

but it's more likely that you will use a compiler to generate your programs.
In general, a compiler translates programs from a source language
to a target language.
Compilers whose target language is WebAssembly may take
Rust, C, Go, or a variety of other languages as a source language.
In this case, the compiler produces a WebAssembly _core module_.
Core modules have some limitations, which components address.

## Core modules

Typically, a core module defines a set of _functions_
along with auxiliary definitions
that are necessary for executing those functions.
Functions are made up of _instructions_.

> Auxiliary definitions include:

Author: Changed in [0c2edc5](https://github.com/bytecodealliance/component-docs/pull/281/commits/0c2edc503abc6131618aa014e8b717ca229dfd71)

> * _Linear memories_ define untyped buffers that can be read from

Author: Changed in [0c2edc5](https://github.com/bytecodealliance/component-docs/pull/281/commits/0c2edc503abc6131618aa014e8b717ca229dfd71)

Good point, "untyped" wasn't quite right. Changed to "buffers of uninterpreted bytes".

> * _Linear memories_ define untyped buffers that can be read from

Reviewer: Thanks, that's perfect!

  and written to by instructions.
* _Imports_ define the names of other modules
   that are required to be available to execute
   the functions in the module,
   along with type signatures for required functions
   in the imported module.
* _Exports_ define the names of functions within
  the module that should be accessible externally.
* And others; see [the Core Specification](https://webassembly.github.io/spec/core/syntax/modules.html)
  for the complete list.

> A core module usually corresponds to a single `.wasm` file.

Author: Changed to "binary `.wasm` file" to cover all bases.

I'm not sure what you mean by example output -- do you mean showing the binary contents of a `.wasm` file or the text-based representation of it that a tool would print out?

> A core module usually corresponds to a single `.wasm` file.

Author: Ah, got it. Let me know if this is what you had in mind: [873fb8e](https://github.com/bytecodealliance/component-docs/pull/281/commits/873fb8ee1d48c7fae51629b2aee4892068e75002)

> A core module usually corresponds to a single `.wasm` file.

Reviewer: Yeah that's perfect -- I'm not sure the modern developer really uses `file` as much as in the past, but definitely nice to at least leave that for those who do/can!

These modules can be run in the browser,

> or via a separate runtime such as [Wasmtime](https://wasmtime.dev/)

Reviewer 2: "server-side" might be too specific, since Wasmtime/WAMR are used in various client and embedded devices too.

> or via a separate runtime such as [Wasmtime](https://wasmtime.dev/)

Reviewer: Good point! I'll resolve this then and we can avoid making this change

or [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime).

A module is defined by the [WebAssembly Core Specification](https://webassembly.github.io/spec/core/).

### Limitations of core modules

Core modules are, however, limited in how they expose their functionality to the outside world.

> Core modules expose their functionality by exporting functions.

Reviewer: ```suggestion
Core modules expose their functionality by exporting functions.
```
I don't think this changes with respect to components, so maybe not worth mentioning? Or was the idea to contrast this with interfaces exported w/ the component model?

Author: Good point, I rephrased in [17b64a4](https://github.com/bytecodealliance/component-docs/pull/281/commits/17b64a492b6292100a9ebbfb0b7df0d203f4eb4a).

> These functions' argument and return types are restricted, essentially,

Author: Changed in [17b64a4](https://github.com/bytecodealliance/component-docs/pull/281/commits/17b64a492b6292100a9ebbfb0b7df0d203f4eb4a) (slightly different phrasing)

> to only integers and floating-point numbers.

Author: Changed in [17b64a4](https://github.com/bytecodealliance/component-docs/pull/281/commits/17b64a492b6292100a9ebbfb0b7df0d203f4eb4a) (slightly different phrasing)

Compound types, such as strings, lists, arrays, records, or structs,

> have to be represented in terms of integers and floating-point numbers.

Author: Changed in [17b64a4](https://github.com/bytecodealliance/component-docs/pull/281/commits/17b64a492b6292100a9ebbfb0b7df0d203f4eb4a) (slightly different phrasing)

Recall that a linear memory is an uninitialized region of bytes
declared within a module.
So, a string argument might be represented as two separate arguments:

> an integer offset into a memory,

Author: Good idea, added in [17b64a4](https://github.com/bytecodealliance/component-docs/pull/281/commits/17b64a492b6292100a9ebbfb0b7df0d203f4eb4a)

and an integer representing the length of the string.
These representations are frequently specific to each programming language.
For example, a string in C might be represented entirely differently
from a string in Rust or in JavaScript.
Moreover, to make this approach work, modules must import and export memories,
which can be error-prone, as different languages
make different assumptions about memory layout.

For WebAssembly modules to interoperate, therefore, there needs to be an agreed-upon way
to expose these richer types across module boundaries.

## Components

> Components were developed to ease interoperability between modules.

Author: See what you think of [5e65915](https://github.com/bytecodealliance/component-docs/pull/281/commits/5e65915ed87c1d8acd3aef90895fb9f6dcb81a09)?

> Components were developed to ease interoperability between modules.

Reviewer: I think it's good! I hesitated to start the section with a back-reference (in case someone skipped around), but I think it definitely lays out those two benefits well. If someone stopped reading after that sentence, they'd at least pull the right intuition away.

Conceptually, a component is a module that is restricted
to interact only through its imported and exported functions.
Compared to core modules, components also use a richer
mechanism for expressing the types of functions.

### Interfaces

> These interfaces are expressed in a separate language called [WIT (Wasm Interface Type)](./wit.md).

Author: Hmm, I've preferred using "WebAssembly" over "Wasm" for consistency, but in https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md , WIT is introduced as "The Wasm Interface Type (WIT) format". Should it be different here?

> These interfaces are expressed in a separate language called [WIT (Wasm Interface Type)](./wit.md).

Reviewer: Thanks this is a good point -- Wasm is a shortening of WebAssembly so I'm not sure it matters a ton -- for example [`wit-bindgen` uses "WebAssembly Interface Types"](https://github.com/bytecodealliance/wit-bindgen), I *think* the words should be interchangable.

That said, consistency is better, so I'm happy with Wasm Interface Types! What I really wanted to point out was to have the acronym in parens and the full phrase first, at least that's my personal preference

> These interfaces are expressed in a separate language called [WIT (Wasm Interface Type)](./wit.md).

Author: I'm fine with using "WebAssembly" here if both are used elsewhere, so I changed it to what you suggested ( [117804e](https://github.com/bytecodealliance/component-docs/pull/281/commits/117804e6f61c79c1def68d165d566b276f615329) ).

> Interfaces contain _types_.

Reviewer: types, and functions, maybe worth linking directly to [the WIT section](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#wit-interfaces) here, or to our WIT explainer

Author: Changed in [005f1a6](https://github.com/bytecodealliance/component-docs/pull/281/commits/005f1a652211cdbd397759678732a0c4d15ffa24)

The bit-level representations of these types are specified by
the [Canonical ABI (Application Binary Interface)](./../advanced/canonical-abi.md).

### Interoperability

Interfaces make it possible to write components that are

> portable across different architectures and operating systems.

Reviewer: I'm not sure this is true -- it's more that they enable bundling different types and functionality in high level languages which can be used across components. 

I think WebAssembly core accomplishes these goals inherently

Not only that, components are portable across different programming languages.
A component implemented in Go can communicate directly and safely

> with a C or Rust component.

Reviewer: ```suggestion
with a C or Rust component, by relying on the shared conventions of the Component Model ABI.
```

Writing a component doesn't even require knowledge
of which language its dependent components are implemented in,
only the component interface expressed in WIT.

> Additionally, components can be linked into larger graphs,
> with one component's exports satisfying another's imports.

Reviewer: ```suggestion
Additionally, components can be composed into larger graphs,
with one component's exports satisfying another's imports.
```

Good time to link to https://component-model.bytecodealliance.org/composing-and-distributing/composing.html IMO

### Benefits of the component model

Putting all of the pieces together:
the component model is a way of writing WebAssembly modules
that interact with each other only through exports and imports of functions
whose types are expressed using WIT.

> When combined with Wasm's strong [sandboxing](https://webassembly.org/docs/security/),

Reviewer: I think this wording might be a little weird, because the component model can't exist *without* Wasm as the lower layer essentially... So more like "building upon" than "combined with"?

Author: Changed in [262cec0](https://github.com/bytecodealliance/component-docs/pull/281/commits/262cec0552e319c517020d165626db81dc9dcd08)

the component model has further benefits.
Richer type signatures express richer semantic properties
than type signatures made up only of integers and floats.
Rich types make it possible to statically analyse
and reason about a component's behaviour.
Simply by examining the surface of the component—the types
of its imports and exports—properties can be

> enforced and guaranteed.

Reviewer: Do you think it might be a good idea to put an example WIT here to drive the point home? Might help to also break up the prose here

Author: I think including WIT might be jumping ahead -- there would need to be a lot of explanation of the syntax that doesn't seem to fit into a persuasive mode. A diagram might help; I'm not the greatest with diagrams, but maybe I can think about that after the text is in agreed-upon shape.

The relationships within a graph of components can be analysed:
for example, to verify that a component containing business logic
has no access to a component containing personally identifiable information.

Moreover, a component interacts with a runtime or other components
_only_ by calling its imports and having its exports called.
Specifically, unlike core modules, a component may not export a memory
and thus it cannot indirectly communicate to others
by writing to its memory and having others read from that memory.
This not only reinforces sandboxing, but enables interoperation
between languages that make different assumptions about memory:
for example, allowing a component that relies garbage-collected memory
to interoperate with one that uses conventional linear memory.

## Using components

Now that you have a better idea about how the component model can help you, take a look at [how to build components](../language-support.md) in your favorite language!

## Further reading

> 

Reviewer: Now that this is a whole section, we can probably remove the `>` ?

I'm actually not even sure why it's a quote like that -- probably was intended to highlight?

Author: Fixed in [262cec0](https://github.com/bytecodealliance/component-docs/pull/281/commits/262cec0552e319c517020d165626db81dc9dcd08)

# Why the Component Model?

At a high level, the component model builds upon WebAssembly _core modules_

> to enhance interoperability, both by enriching the type system

Reviewer 2: ```suggestion
to enhance interoperability between languages and libraries, both by enriching the type system
```

used for checking the safety of interactions between modules,

> and by removing potentially error-prone ways for modules to interact.

Reviewer 2: ```suggestion
and by clearly defining and enforcing the low-level calling contract between separately-compiled modules.
```

To understand what the limitations of core modules are,
we start by defining them.

## WebAssembly Core modules

A module is defined by the [WebAssembly Core Specification](https://webassembly.github.io/spec/core/).

WebAssembly programs can be written by hand,

> but it's more likely that you will use a compiler to generate your programs.
> In general, a compiler translates programs from a source language
> to a target language.
> Compilers whose target language is WebAssembly may take
> Rust, C, Go, or a variety of other languages as a source language.

Reviewer: ```suggestion
but it's more likely that you will use a higher level programming language
such as Rust, C, Go, Javascript, or Python to build WebAssembly programs.
```

What do you think about this? I figured we didn't need to try and explain compilers, and that actually paradoxically feels like leaving out interpreted languages we support.

> In this case, the compiler produces a WebAssembly [core module](https://webassembly.github.io/spec/core/syntax/modules.html).

Reviewer: ```suggestion
Many existing toolchains currently produce a [WebAssembly core module](https://webassembly.github.io/spec/core/syntax/modules.html) -- a single binary `.wasm` file.
```

A core module is a set of definitions.
Kinds of definitions include:
* _Functions_ define executable units of code
  (sequences of instructions along with declarations
  for argument names and types and return types).
* [_Linear memories_](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-mem)
  define buffers of uninterpreted bytes that can be read from
  and written to by instructions.
* _Imports_ define the names of other modules
   that are required to be available to execute
   the functions in the module,
   along with type signatures for required functions
   in the imported module.
* _Exports_ define the names of functions within
  the module that should be accessible externally.
* And others; see [the Core Specification](https://webassembly.github.io/spec/core/syntax/modules.html)
  for the complete list.

> A core module usually corresponds to a single binary `.wasm` file.

Reviewer: ```suggestion
```

This line may not be needed if you move it up above (earlier comment)

These modules can be run in the browser,
or via a separate runtime such as [Wasmtime](https://wasmtime.dev/)
or [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime).

### Limitations of core modules

> Core modules are, however, limited in how they expose their functionality to the outside world.

Reviewer: ```suggestion
Core modules are limited in the computation they can perform and 
how they expose their functionality to the outside world.
```

In WebAssembly core modules, functions are restricted, essentially,
to using integer (`i32` or `i64`) or floating-point (`f32` or `f64`) types.
Only these types can be passed as arguments to functions,
and only these types can be returned from functions as results.
Compound types common in higher-level programming languages,
such as strings, lists, arrays, enums (enumerations), or structs (records),
have to be represented in terms of integers and floating-point numbers.

> Recall that a linear memory is an uninitialized region of bytes

Reviewer 2: ```suggestion
Recall that a linear memory is an uninterpreted array of bytes
```

Author: Already fixed, see the current diff :)

declared within a module.
So, a string argument might be represented as two separate arguments:
an integer offset into a memory,

> and an integer representing the length of the string.

Reviewer: What do you think about swapping these sentences and starting with "For example"

So something like:

> For example, for a function to accept a string, ... into a linear memory.
> Recall that a linear memory is an ...

In pseudocode, a type signature for a string-manipulating function
might look like:

```
remove-duplicates: func(offset: i32, length: i32) -> [i32, i32]
```

supposing that `remove-duplicates` is a function
to create a new string consisting of the unique characters
in its argument.
The return type is a pair of 32-bit integers,
reflecting that the function must return
the new offset for the newly allocated string, as well as its length.

For this to work, the module defining the `remove-duplicates` function
would also need to include
an export declaration that exports a memory to be used
for the argument and result strings. Pseudocode:

```
export "string_mem" (mem 1)
```

And, the module using the `remove-duplicates` function
would need to import this memory. Pseudocode:

```
import "strings" "string_mem"
```

(This pseudocode is still simplified, since the importer
also needs to declare the size of the memory being
imported.)

We would prefer to write a pseudocode type signature like this:

```
remove-duplicates: func(s: string) -> string
```

and dispense with the memory exports and imports altogether.

> Data representations are frequently specific to each programming language.

Reviewer: We could maybe enhance this just slightly with a phrase like "The complexity doesn't stop there", because just getting the bytes over is one thing, properly interpreting them in the receiver is another!

Author: Also a good idea -- done in [bfe929b](https://github.com/bytecodealliance/component-docs/pull/281/commits/bfe929b6e611c4459b712d96343ebc13c35e3ecd)

For example, a string in C is represented entirely differently
from a string in Rust or in JavaScript.
Moreover, to make this approach work, modules must import and export memories,
which can be error-prone, as different languages
make different assumptions about memory layout.

For WebAssembly modules written in different languages to interoperate smoothly,
there needs to be an agreed-upon way to expose these richer types across module boundaries.

## Components

Components solve the two problems that we've seen so far:
the limited type system of core module functions,
and cross-language interoperability.

> Conceptually, a component is a module that is restricted

Reviewer: ```suggestion
Conceptually, a component is a WebAssembly binary (which may or may not contain modules) that is restricted
```

Author: Adopted the suggestion, but it's more accurate to say that the binary interacts through the contained _modules'_ imports and exports, right?

Reviewer: So that's kind of true, but I think it's not unreasonable to think about it it as interacting via the component model itself. A component wraps modules, and sort of... intermediates their inputs and outputs -- so it's a bit murky. 

It's also *technically* possible to have a component that does *not* have an internal module actually. For example there are sometimes... patch/adapter subcomponents (not modules IIRC) that exist to patch imports to exports and do some other very internal things.

to interact only through its imported and exported functions.

> Compared to core modules, components also use a richer
> mechanism for expressing the types of functions.

Reviewer: ```suggestion
Compared to core modules, components use a different binary format and opt-in to a richer
mechanism for expressing the types of functions by default.
```

Author: Adopted suggestion (with a little editing) -- I thought about including the `file` output here to contrast with the earlier example of `file` output, but it doesn't seem too helpful, since it doesn't explicitly mention components:

```
$ file add.component.wasm 
add.component.wasm: WebAssembly (wasm) binary module version 0x1000d
```

What do you think?

Reviewer: Ah, so here, my thought was that the *contrast* is how you know it's a Component. Regular modules will output the other message (with "MVP") and the other code. 

Honestly one of the easiest ways most people do it is to run `wasm-tools print <file> | less` and check if `(component` is at the top...

Reviewer: OK I really want to get this in so let's add this in a follow up if we want to -- IMO it's worth adding but I could definitely be convinced otherwise

### Interfaces

These interfaces are expressed in a separate language called [WIT (Wasm Interface Type)](./wit.md).
[Interfaces](./wit.md#interfaces) contain definitions of _types_
and type signatures for [_functions_](./wit.md#functions).
The bit-level representations of types are specified by
the [Canonical ABI (Application Binary Interface)](./../advanced/canonical-abi.md).

### Interoperability

WebAssembly core modules are already portable across different architectures
and operating systems;

> components retain these benefits and, using interfaces,

Reviewer: ```suggestion
components retain these benefits and, using the Component Model ABI,
```

add portability across different programming languages.
A component implemented in Go can communicate directly and safely
with a C or Rust component, by relying on the shared conventions of the Component Model ABI.
Writing a component doesn't even require knowledge
of which language its dependent components are implemented in,
only the component interface expressed in WIT.
Additionally, components can be [composed](../composing-and-distributing.md) into larger graphs,
with one component's exports satisfying another's imports.

> ### Benefits of the component model

Reviewer: Do we want "Component Model" in most places?

Author: Do you mean whether to capitalize "Component Model" or not? I have no strong feelings as long as it's consistent. (Or do you mean whether to say "component model" or "components"? Since the section is headed 'why the component model?' I'm more inclined to use 'component model' at least in this section...)

Putting all of the pieces together:

> the component model is a way of writing WebAssembly modules
> that interact with each other only through exports and imports of functions
> whose types are expressed using WIT.

Reviewer: ```suggestion
the component model is a binary WebAssembly format encapsulates WebAssembly
modules and enables interactions across modules through exports and imports of functions
whose types are expressed using WIT.
```

Author: Did this a little differently in [078c562](https://github.com/bytecodealliance/component-docs/pull/281/commits/078c5622bc7dcce3faca191b4a345b4452a17dd3) (mainly, I didn't want to say that the component model _is_ the format, since (at least to me) the "model" is a way of thinking and the concrete artifacts like the binary format reflect that way of thinking, or something like that.)

Building upon Wasm's strong [sandboxing](https://webassembly.org/docs/security/),
the component model has further benefits.
Richer type signatures express richer semantic properties
than type signatures made up only of integers and floats.

> Rich types make it possible to statically analyse
> and reason about a component's behaviour.
> Simply by examining the surface of the component—the types
> of its imports and exports—properties can be

Reviewer 2: ```suggestion
Rich types make it easier to know what a component or interface is 
doing at a glance and have guarantees of what bad things cannot happen.
```

A core module usually corresponds to a single binary `.wasm` file.
These modules can be run in the browser,
or via a separate runtime such as [Wasmtime](https://wasmtime.dev/)
or [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime).

### Limitations of core modules

Core modules are, however, limited in how they expose their functionality to the outside world.
In WebAssembly core modules, functions are restricted, essentially,
to using integer (`i32` or `i64`) or floating-point (`f32` or `f64`) types.
Only these types can be passed as arguments to functions,
and only these types can be returned from functions as results.
Compound types common in higher-level programming languages,
such as strings, lists, arrays, enums (enumerations), or structs (records),
have to be represented in terms of integers and floating-point numbers.

Recall that a linear memory is an uninitialized region of bytes
declared within a module.
So, a string argument might be represented as two separate arguments:
an integer offset into a memory,
and an integer representing the length of the string.

In pseudocode, a type signature for a string-manipulating function
might look like:

```

> remove-duplicates: func(offset: i32, length: i32) -> [i32, i32]

Reviewer: You know what, this is fine -- I think it's easy to parse/grok

```

supposing that `remove-duplicates` is a function
to create a new string consisting of the unique characters
in its argument.
The return type is a pair of 32-bit integers,
reflecting that the function must return

> the new offset for the newly allocated string, as well as its length.

Reviewer: This is a *great* example. I'm not sure what I expected, maybe string reverse or something, but this is pretty useful seeming/realistic.

The only thing I'd add here is that we should explain the args as skillfully as we explain the return type. I know it's down in the weeds but I think actually showing how painful it is will motivate even more effectively.

Right now for the return argument we have this:

> The return type is a pair of 32-bit integers,
> reflecting that the function must return
> the new offset for the newly allocated string, as well as its length.

Maybe this could be something like:

> The return type is a list of integers, the first being an offset into the linear memory of the
> component -- where the newly allocated string starts -- along with a length.
> After calling this function, the caller has to reach into the linear memory and attempt to read
> the created string out, given the provided offset and length.

Author: Thanks! Yes, I was going to use string reverse, but realized it's more motivating to use an example where the output string length can be different from the input length.
